Mesh_3/include/CGAL/Mesh_3/tet_soup_to_c3t3.h

```diff
@@ -0,0 +1,447 @@
+void add_facet_to_incident_cells_map(const typename Tr::Cell_handle c, int i,
+                                     std::map<std::set<typename Tr::Vertex_handle>,
+                                     std::vector<std::pair<typename Tr::Cell_handle,
+                                     int> > >& incident_cells_map)
+{
+  typedef typename Tr::Vertex_handle                            Vertex_handle;
+  typedef typename Tr::Cell_handle                              Cell_handle;
+  typedef std::set<Vertex_handle>                               Facet;
+  typedef std::pair<Cell_handle, int>                           Incident_cell;
+  typedef std::map<Facet, std::vector<Incident_cell> >          Incident_cells_map;
+
+  // the opposite vertex of f in c is i
+  Facet f;
+  f.insert(c->vertex((i + 1) % 4));
+  f.insert(c->vertex((i + 2) % 4));
+  f.insert(c->vertex((i + 3) % 4));
+  CGAL_precondition(f.size() == 3);
+
+  Incident_cell e = std::make_pair(c, i);
+  std::vector<Incident_cell> vec;
+  vec.push_back(e);
+  std::pair<typename Incident_cells_map::iterator, bool> is_insert_successful =
+      incident_cells_map.insert(std::make_pair(f, vec));
+  if(!is_insert_successful.second) // the entry already exists in the map
+  {
+    // a facet must have exactly two incident cells
+    CGAL_assertion(is_insert_successful.first->second.size() == 1);
+    is_insert_successful.first->second.push_back(e);
+  }
+}
+
+template<class Tr>
+void build_finite_cells(Tr& tr,
+                        const std::vector<boost::array<int,5> >& finite_cells,
+                        const std::vector<typename Tr::Vertex_handle>& vertex_handle_vector,
+                        std::map<std::set<typename Tr::Vertex_handle>,
+                        std::vector<std::pair<typename Tr::Cell_handle,
+                        int> > >& incident_cells_map,
+                        const std::map<boost::array<int,3>, int>& border_facets)
+{
+  typedef boost::array<int, 5>              Tet_with_ref; // 4 ids + 1 reference
+
+  typedef typename Tr::Vertex_handle                            Vertex_handle;
+  typedef typename Tr::Cell_handle                              Cell_handle;
+
+  // build the finite cells
+  for(std::size_t i=0; i<finite_cells.size(); ++i)
+  {
+    const Tet_with_ref& tet = finite_cells[i];
+    boost::array<Vertex_handle, 4> vs;
+
+    for(int j=0; j<4; ++j)
+    {
+      CGAL_precondition(static_cast<std::size_t>(tet[j]) < tr.number_of_vertices() &&
+                        tet[j] >= 0);
+      vs[j] = vertex_handle_vector[tet[j] + 1];
+      CGAL_postcondition(vs[j] != Vertex_handle());
+    }
+
+    // this assertion also tests for degeneracy
+    CGAL_assertion(CGAL::orientation(vs[0]->point().point(), vs[1]->point().point(),
+        vs[2]->point().point(), vs[3]->point().point()) == POSITIVE);
+
+    Cell_handle c = tr.tds().create_cell(vs[0], vs[1], vs[2], vs[3]);
+    c->info() = tet[4]; // the reference encodes the interior/exterior info
```

Comment:
Actually the comment is wrong. In a Medit file, there is no tetrahedron that is "exterior". All the tetrahedron listed in the file are in the mesh. The "reference" of cells is just another name for the "subdomain index". Different software, different wording.